lib/ddtrace/tracer.rb

```diff
@@ -446,4 +465,78 @@ def deactivate_priority_sampling!(base_sampler = nil)
       :record_context,
       :write
   end
+
+  # Represents the act of taking a span measurement.
+  # It gives a Span a context which can be used to
+  # manage and decorate the Span.
+  # When completed, it yields the Span.
+  class SpanOperation
+    extend Forwardable
+
+    INCLUDED_METHODS = [:==].to_set.freeze
+    EXCLUDED_METHODS = [:finish, :parent, :parent=].to_set.freeze
+
+    def initialize(span_name, options = {})
+      # Resolve service name
+      parent = options[:child_of]
+      options[:service] ||= parent.service unless parent.nil?
+
+      # Build span
+      @span = Span.new(
+        span_name,
+        options
+      )
+      @tracer = options[:tracer]
+      @context = options[:context]
+
+      # Add span to the context, if provided.
+      @context.add_span(self) if @context
+
+      if parent.nil?
+        # Root span: set default tags.
+        set_tag(Datadog::Ext::Runtime::TAG_PID, Process.pid)
+        set_tag(Datadog::Ext::Runtime::TAG_ID, Datadog::Core::Environment::Identity.id)
+      else
+        # Only set parent if explicitly provided.
+        # We don't want it to override context-derived
+        # IDs if it's a distributed trace w/o a parent span.
+        self.parent = parent
+      end
+
+      # Set tags if provided.
+      set_tags(options[:tags]) if options.key?(:tags)
+    end
+
+    attr_reader :parent
+    attr_accessor :span, :tracer, :context
+
+    # Set span parent
+    def parent=(parent)
+      @parent = parent
+      span.parent = parent && parent.span
+    end
+
+    def finish(end_time = nil)
+      return span if finished?
+
+      # Stop the span
+      span.stop(end_time)
+
+      begin
+        context.close_span(self) if context
+        tracer.record_span(self) if tracer
+      rescue StandardError => e
+        Datadog.logger.debug("error recording finished trace: #{e} Backtrace: #{e.backtrace.first(3)}")
+        Datadog.health_metrics.error_span_finish(1, tags: ["error:#{e.class.name}"])
+      end
+
+      span
+    end
+
+    # Forward instance methods except ones that would cause identity issues
+    def_delegators :span, *(Span.instance_methods(false).to_set - EXCLUDED_METHODS)
```

Comment:
This somewhat makes it hard to understand this class without understanding the `Span` class. I would suggest a similar approach that is more deliberate: actually have a `FORWARDED_METHODS` constant that lists the methods we want to forward BUT put in a spec that this constant must be the same as  `Span.instance_methods(false).to_set - EXCLUDED_METHODS`.

That way we're forced to remember that a change to `Span` may need a change to this class BUT we're also forced to revisit that decision -- maybe whatever we just added to `Span` shouldn't be forwarded automatically.

Reply:
> actually have a FORWARDED_METHODS constant that lists the methods we want to forward

This is how I had it before, but I didn't like the drift I was getting, as you allude to the change part. Perhaps asserting this in the spec is a reasonable way to handle that, so we can be explicit again.

What's the biggest disadvantage to this being implicit? Is it just clarity?

Reply:
I'd say it not only increases clarity, but also makes the API for this class explicit and changes deliberate.

For the clarity part, having this information here allows one to read and understand what this class does by itself without needing to reference the `Span` class. (I believe, but haven't checked, that it would also enable sorbet to figure it out as well.)

For the explicit, consider than in the future we add some operation on `Span` that doesn't make sense to have while the span is active (e.g. while there is a `SpanOperation` attached to it). In the pull request, for that feature, there would need to be a diff to add it to `FORWARDED_METHODS` or `EXCLUDED_METHODS`, and we could validate that addition clearly.

With the current implicit forwarding, we could add `:finish_foo` to `Span`, and it would automatically become part of the `SpanOperation` API, and we might not remember this and later find out "oops, this shouldn't have happened".

Reply:
Okay, I don't think its too crazy to add explicit forwarding if it's a design improvement. I'll give this a go.

Reply:
This seemed to work just fine. Will add specs for it as well.